### PR TITLE
refactor(gossipsub)!: initialize `ProtocolConfig` from `GossipsubConfig`

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

This simplifies the tests as we don't have to go through the `new_handler` abstraction.

## Notes

Extracted out of https://github.com/libp2p/rust-libp2p/pull/3254.
Unfortunately, this is a breaking change because the `protocol` module is public. See https://github.com/libp2p/rust-libp2p/pull/3303#discussion_r1084698192 for more discussion.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/gossipsub/CHANGELOG.md [hunk 1]
```diff
@@ -6,7 +6,10 @@
 
 - Update to `libp2p-swarm` `v0.42.0`.
 
+- Initialize `ProtocolConfig` via `GossipsubConfig`. See [PR 3381].
+
 [PR 3207]: https://github.com/libp2p/rust-libp2p/pull/3207/
+[PR 3381]: https://github.com/libp2p/rust-libp2p/pull/3381/
```
Comment (mxinden) [inline protocols/gossipsub/CHANGELOG.md:12]: Why does this need a changelog entry? This is not user facing, right?

**Review by mxinden — APPROVED** [review]
Other than the changelog entry, good to merge from my end.

File: protocols/gossipsub/CHANGELOG.md (hunk 1 above)
Comment (thomaseizinger) [reply protocols/gossipsub/CHANGELOG.md:12]: It is (unfortunately) part of the public API.

Comment (thomaseizinger) [reply protocols/gossipsub/CHANGELOG.md:12]: Happy to remove it again if you think that doesn't warrant an entry!

Comment (mxinden) [reply protocols/gossipsub/CHANGELOG.md:12]: Oh, I am surprised. No strong opinion. Let's keep as is :+1:

Comment (thomaseizinger) [reply protocols/gossipsub/CHANGELOG.md:12]: Yep, it's something I'd like to fix. See linked discussion in PR description.

**Comment by mxinden** [comment]
Clippy stable is continuously failing with:

```
 error[E0061]: this function takes 1 argument but 5 arguments were supplied
Error:    --> protocols/gossipsub/src/behaviour/tests.rs:275:17
    |
275 |                 ProtocolConfig::new(Cow::from(""), None, 0, ValidationMode::None, false),
    |                 ^^^^^^^^^^^^^^^^^^^                ----  -  --------------------  ----- argument of type `bool` unexpected
    |                                                    |     |  |
    |                                                    |     |  argument of type `config::ValidationMode` unexpected
    |                                                    |     argument of type `{integer}` unexpected
    |                                                    argument of type `std::option::Option<_>` unexpected
    |
note: expected `&config::GossipsubConfig`, found enum `std::borrow::Cow`
   --> protocols/gossipsub/src/behaviour/tests.rs:275:37
    |
275 |                 ProtocolConfig::new(Cow::from(""), None, 0, ValidationMode::None, false),
    |                                     ^^^^^^^^^^^^^
    = note: expected reference `&config::GossipsubConfig`
                    found enum `std::borrow::Cow<'_, str>`
note: associated function defined here
   --> protocols/gossipsub/src/protocol.rs:60:12
    |
60  |     pub fn new(gossipsub_config: &GossipsubConfig) -> ProtocolConfig {
    |            ^^^ ----------------------------------
help: remove the extra arguments
    |
275 |                 ProtocolConfig::new(/* &config::GossipsubConfig */),
    |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For more information about this error, try `rustc --explain E0061`.
error: could not compile `libp2p-gossipsub` due to previous error
```

**Comment by thomaseizinger** [comment]
Thanks for fixing this, must have been a bad merge or something!

**Pull request merged**